src/libsystemd/sd-daemon/sd-daemon.c

```diff
@@ -467,13 +467,34 @@ _public_ int sd_pid_notify_with_fds(
         if (!e)
                 return 0;
 
-        r = sockaddr_un_set_path(&sockaddr.un, e);
+        /* Allow AF_UNIX and AF_VSOCK, reject the rest. */
+        r = socket_local_address_parse(&address, e);
         if (r < 0)
                 goto finish;
-        msghdr.msg_namelen = r;
+        msghdr.msg_namelen = address.size;
+
+        /* If we didn't get an address, talk to the hypervisor by default, as that's the purpose of this functionality. */
+        if (address.sockaddr.vm.svm_family == AF_VSOCK && address.sockaddr.vm.svm_cid == VMADDR_CID_ANY)
+                address.sockaddr.vm.svm_cid = VMADDR_CID_HYPERVISOR;
+
+        /* At the time of writing QEMU does not yet support AF_VSOCK + SOCK_DGRAM and returns
+         * ENODEV. Fallback to SOCK_SEQPACKET in that case. */
+        fd = socket(address.sockaddr.sa.sa_family, SOCK_DGRAM|SOCK_CLOEXEC, 0);
+        if (fd < 0 && errno == ENODEV && address.sockaddr.sa.sa_family == AF_VSOCK) {
+                fd = socket(address.sockaddr.sa.sa_family, SOCK_SEQPACKET|SOCK_CLOEXEC, 0);
+                if (fd < 0) {
+                        r = -errno;
+                        goto finish;
+                }
+
+                if (connect(fd, &address.sockaddr.sa, address.size) < 0) {
+                        r = -errno;
+                        goto finish;
+                }
 
-        fd = socket(AF_UNIX, SOCK_DGRAM|SOCK_CLOEXEC, 0);
-        if (fd < 0) {
+                msghdr.msg_name = NULL;
+                msghdr.msg_namelen = 0;
+        } else if (fd < 0) {
```

Comment:
maybe:

```c
if (r < 0) {
        if (!(errno == ENODEV || ERRNO_IS_NOT_SUPPORTED(errno)) || address.sockaddr.sa.sa_family != AF_VSOCK)
               return -errno;
   
        …
```

Reply:
Not sure I follow, in this branch the fd is invalid so we always need to error out

Reply:
I am just saying that you check for `fd < 0` twice. and I'd nest that: check for it once, and then check if we are in the fallback condition, and if not exit.

Reply:
ah ok, got it